# 2. إنشاء الليبل وتحديد مشكلة عدم توازن البيانات (Label Creation & Class Imbalance)

In [2]:
import pandas as pd

df = pd.read_csv('olist_master_artifact.csv')
df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,total_price,total_freight,...,review_answer_timestamp,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,29.99,8.72,...,2017-10-12 03:43:48,7c396fd4830fd04220f754e42b4e5bff,3149,sao paulo,SP,3149.0,-23.574809,-46.587471,sao paulo,SP
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,118.70,22.76,...,2018-08-08 18:37:50,af07308b275d755c9edb36a90c618231,47813,barreiras,BA,47813.0,-12.169860,-44.988369,barreiras,BA
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,159.90,19.22,...,2018-08-22 19:07:58,3a653a41f6f9fc3d2a113cf8398680e8,75265,vianopolis,GO,75265.0,-16.746337,-48.514624,vianopolis,GO
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00,45.00,27.20,...,2017-12-05 19:21:58,7c142cf63193a1473d2e66489a9ae977,59296,sao goncalo do amarante,RN,59296.0,-5.767733,-35.275467,sao goncalo do amarante,RN
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00,19.90,8.72,...,2018-02-18 13:02:51,72632f0f9dd73dfee390c9b22eb56dd6,9195,santo andre,SP,9195.0,-23.675037,-46.524784,santo andre,SP


In [6]:
df['order_delivered_customer_date'] = df['order_delivered_customer_date'].astype('datetime64[ns]')
df['order_delivered_customer_date']

0       2017-10-10 21:25:13
1       2018-08-07 15:27:45
2       2018-08-17 18:06:29
3       2017-12-02 00:28:42
4       2018-02-16 18:17:02
                ...        
99436   2017-03-17 15:08:01
99437   2018-02-28 17:37:56
99438   2017-09-21 11:24:17
99439   2018-01-25 23:32:54
99440   2018-03-16 13:08:30
Name: order_delivered_customer_date, Length: 99441, dtype: datetime64[ns]

In [7]:
df['order_estimated_delivery_date'] = df['order_estimated_delivery_date'].astype('datetime64[ns]')
df['order_estimated_delivery_date']

0       2017-10-18
1       2018-08-13
2       2018-09-04
3       2017-12-15
4       2018-02-26
           ...    
99436   2017-03-28
99437   2018-03-02
99438   2017-09-27
99439   2018-02-15
99440   2018-04-03
Name: order_estimated_delivery_date, Length: 99441, dtype: datetime64[ns]

In [8]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  object        
 4   order_approved_at              99281 non-null  object        
 5   order_delivered_carrier_date   97658 non-null  object        
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
 8   total_price                    98666 non-null  float64       
 9   total_freight                  98666 non-null  float64       
 10  total_items                    98666 non-null  float64       
 11  total_weight_g 

In [13]:
df['is_late'] = (df['order_delivered_customer_date'] > df['order_estimated_delivery_date']).astype(int)

In [ ]:
# 0 the number of on-time deliveries
# 1 the number of late orders
df['is_late'].value_counts()

is_late
0    91614
1     7827
Name: count, dtype: int64

In [15]:
# حفظ الداتا مع الليبل الجديد بملف CSV
df.to_csv('olist_with_labels.csv', index=False)

print("Successfully saved olist_with_labels.csv!")

Successfully saved olist_with_labels.csv!


# Class Imbalance Observation:
### The label distribution shows a significant class imbalance; on-time deliveries (0) account for approximately 92% of the dataset (91,614 orders), while delayed deliveries (1) account for only 8% (7,827 orders). This imbalance means that relying solely on Accuracy for model evaluation will be misleading, and we should consider using appropriate metrics (like Precision, Recall, and F1-Score) or handling techniques (such as class_weight) during the machine learning phase.